### PR TITLE
DRAFT - SIP on Netty

### PR DESCRIPTION
**Pull request by wtlucy**
For #15779 

### High Level Changes:
- sipcontainer classes relying on the channel framework have been reimplemented on top of Netty
    - `WsByteBuffer` replaced for[by] `ByteBuf`
    - Channel Framework service and APIs replaced by Liberty Netty framework `io.openliberty.netty.NettyFramework` and `io.netty.*` APIs
    -  classes previously relying on channel framework functionality have been moved into new packages `*.chfw.*` 
        - e.g. `com/ibm/ws/sip/channel/resolver/dns/impl/Dns.java` -> `com/ibm/ws/sip/channel/resolver/dns/impl/chfw/Dns.java`
        - new Netty implementations are in `*.netty.*` packages, e.g. `com/ibm/ws/sip/channel/resolver/dns/impl/netty/Dns.java`
    - a new server.xml config element `<sipContainer useNettyTransport="true/false">` has been added, which toggles between the chfw and netty implementations
    - A few improvements needed to be made to the Liberty Netty framework:
        - support for outbound connection creation via `NettyFramework.startOutbound()`
        - TLS support via a new `io.openliberty.netty.internal.tls.impl`
            - this bundle provides a new service `NettyTlsProvider`, which provides APIs to build a `io.netty.handler.ssl.SslContext` from the active Liberty TLS configs, via `com.ibm.websphere.ssl.JSSEHelper`. 
            - The `SslContext` can then be used to create Netty `SslHandler`

#### DNS resolver (domainResolver)
- Netty changes mostly contianed to `com.ibm.ws.sip.channel.resolver.dns.impl.netty` and `com.ibm.ws.sip.channel.resolver.impl.netty`
    - `com.ibm.ws.sip.channel.resolver.dns.impl.netty` largely contains `WsByteBuffer` -> `ByteBuf` changes
    - `com.ibm.ws.sip.channel.resolver.impl.netty` makes outbound TCP and UDP requests via `SipResolverTcpTransport` and `SipResolverUdpTransport`
        - those impls use `NettyFramework` to create their outbound connection bootstraps, and then `NettyFramework.startOutbound()` to start and register the connections
        - each impl provides a `ChannelInitializerWrapper` which sets up their pipelines - in both cases including a `SimpleChannelInboundHandler` which is used to pass new data to the correct methods
- selection between chfw and Netty implementations occurs in `com.ibm.ws.sip.container.resolver.DomainResolverImpl.getSipURILookup(SipURILookupCallbackImpl, SIPUri)` and `com.ibm.ws.sip.container.resolver.DomainResolverImpl.init()`

#### SIP Endpoint (sipEndpoint)
- Netty changes split between `com.ibm.ws.sip.stack.transport.netty` and `com.ibm.ws.sip.stack.transport.sip.netty`
- `com.ibm.ws.sip.stack.transport` refactored rather than split completely (this might want to be revisited)
- `com.ibm.ws.sip.stack.transport.GenericEndpointImpl` is the `@Component` which registers and manages endpoints
    - Netty and chfw impls are determined in `activate()`
    - previous chfw chain start logic is largely reused by Netty impl, including naming and config props 
    - endpoints started via `GenericTCPChain` and `GenericUDPChain` `update()`
        - bootstraps created by passing endpoint config to `NettyFramework.create*Bootstrap()`
        - registered and started via `NettyFramework.start()`
        - TLS endpoints covered by `GenericTCPChain`, and implemented by adding a `io.netty.handler.ssl.SslHandler.SslHandler` to the head of the `ChannelInitializer` pipeline

#### SIP Endpoint (sipEndpoint)
- Netty changes split between `com.ibm.ws.sip.stack.transport.netty` and `com.ibm.ws.sip.stack.transport.sip.netty`
- `com.ibm.ws.sip.stack.transport` refactored rather than split completely (this might want to be revisited)
- `com.ibm.ws.sip.stack.transport.GenericEndpointImpl` is the `@Component` which registers and manages endpoints
    - Netty and chfw impls are determined in `activate()`
    - previous chfw chain start logic is largely reused by Netty impl, including naming and config props 
    - endpoints started via `GenericTCPChain` and `GenericUDPChain` `update()`
        - bootstraps created by passing endpoint config to `NettyFramework.create*Bootstrap()`
        - registered and started via `NettyFramework.start()`
        - TLS endpoints covered by `GenericTCPChain`, and implemented by adding a `io.netty.handler.ssl.SslHandler.SslHandler` to the head of the `ChannelInitializer` pipeline
- `SipOutboundConnLink` used for outbound connections
    - chfw and Netty impls swapped via `com.ibm.ws.sip.stack.transaction.transport.SIPConnectionsModel.initSupportedTransports()`

### Remaining work
- the packaging of these changes likely needs to be re-thought. Currently this PR results in new bundles being added to the runtime by a GA feature, which is the desired end result but is probably not OK for this type of code.
- `com.ibm.ws.sip.stack.transport.sip.netty.SipOutboundConnLink` should be updated to use `NettyFramework` APIs to create a bootstrap and connect.

### CODE REVIEW

**Comment by wtlucy** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.sipcontainer_fat

**Comment by wtlucy** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.sipcontainer_fat

**Comment by wtlucy** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.sipcontainer_fat

**Comment by wtlucy** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.sipcontainer_fat

**Comment by wtlucy** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.sipcontainer_fat

**Comment by wtlucy** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.sipcontainer_fat

**Comment by wtlucy** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.sipcontainer_fat

**Comment by wtlucy** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.sipcontainer_fat

**Comment by wtlucy** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.sipcontainer_fat

**Comment by wtlucy** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.sipcontainer_fat

**Comment by wtlucy** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.sipcontainer_fat

**Comment by wtlucy** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.sipcontainer_fat

**Comment by wtlucy** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.sipcontainer_fat

**Comment by wtlucy** [comment]
The latest build has finally completely passed the sipcontainer_fat bucket!
image

File: dev/com.ibm.ws.sipcontainer/src/com/ibm/ws/sip/channel/resolver/impl/netty/SipResolverTcpTransport.java [hunk 1]
```diff
@@ -0,0 +1,754 @@
...
+                c_logger.traceDebug("SipResolverTcpTransport:writeRequest: WRITE_STATE_IDLE");
+
+            _lengthBuffer.clear();
+            _lengthBuffer.writeShort((short) requestBuffer.readableBytes());
+
+//            System.out.println("writeRequest writableBytes" + requestBuffer.readableBytes());
+
+            _bufferArray[0] = _lengthBuffer;
+            _bufferArray[1] = requestBuffer;
+            ByteBuf newBuf = Unpooled.copiedBuffer(_bufferArray);
```
Comment (hutchig) [inline dev/com.ibm.ws.sipcontainer/src/com/ibm/ws/sip/channel/resolver/impl/netty/SipResolverTcpTransport.java:276]: Bill, was there a glitch with the composite buffer below that was solved by using an unpooled copied buffer above?

File: dev/com.ibm.ws.sipcontainer/src/com/ibm/ws/sip/container/internal/SipContainerComponent.java [hunk 2]
```diff
@@ -166,6 +172,9 @@ protected void modified(Map<String, Object> properties) {
 		if (c_logger.isTraceDebugEnabled())
 			c_logger.traceDebug("SipContainerComponent modified", properties);
 		PropertiesStore.getInstance().getProperties().updateProperties(properties);
+        useNetty = MetatypeUtils.parseBoolean(
```
Comment (hutchig) [inline dev/com.ibm.ws.sipcontainer/src/com/ibm/ws/sip/container/internal/SipContainerComponent.java:175]: .

File: dev/com.ibm.ws.sipcontainer/src/com/ibm/ws/sip/container/resolver/DomainResolverImpl.java [hunk 3]
```diff
@@ -196,15 +233,26 @@ private void init() {
 			dsProps.put(CoreProperties.SIP_RFC3263_DNS_FAILURE_DETECTION_WINDOW_SIZE_INTERVAL_SEC, interval);
 			dsProps.put(CoreProperties.SIP_RFC3263_ADD_TTL, addTTL);
 			dsProps.put(CoreProperties.SIP_DNS_QUERY_TIMEOUT, queryTimeoutDuration);
-		
-			SipResolverService.initialize(dsProps,m_chfw);
+
+			_useNetty = SipContainerComponent.useNetty();
+		     if (_useNetty) {
```
Comment (hutchig) [inline dev/com.ibm.ws.sipcontainer/src/com/ibm/ws/sip/container/resolver/DomainResolverImpl.java:238]: What does '_' prefix mean in this context?

**Comment by isaacrivriv** [comment]
Closing in favor of https://github.com/OpenLiberty/open-liberty/pull/24890

**Pull request closed without merging**